Tasks 1-17 solved

IbrayevRamil opened the pull request.

nkarulin committed:
Create README.md

IbrayevRamil committed:
task1 done

IbrayevRamil committed:
task1 done

IbrayevRamil committed:
task3 done

IbrayevRamil committed:
task2 done

IbrayevRamil committed:
task4 done

IbrayevRamil committed:
task5 done

IbrayevRamil committed:
task6 done

IbrayevRamil committed:
task11 need fix...

PutienkoEN committed:
Gitignore (not shure it should be here, but i don't know how remove it)

Task1Test Edit(added \n)
Task3Test Edit(added foreach + file route + changed data-source + added white space to poem)
Task4Test Added
Task6Test Added
Task7Test Added
Task19Test Added
Car class constructor changed from package-privte to public (for test)

PutienkoEN committed:
Tests:
1
3
4
6
7
8
9
10
12
19
Texts(resources) for test 9.

IbrayevRamil committed:
task10 done

IbrayevRamil committed:
task9 done

IbrayevRamil committed:
task7 done

IbrayevRamil committed:
task11 done(?)

IbrayevRamil committed:
task12 done

IbrayevRamil committed:
task17 done?

IbrayevRamil committed:
task17 done?

IbrayevRamil committed:
task8 done

IbrayevRamil committed:
task1 fixed and done

IbrayevRamil committed:
task1 fixed and done

PutienkoEN committed:
Disabled Task4Test(there 2 tests)

PutienkoEN committed:
Task19 edited data Provider

PutienkoEN committed:
Added Test2, Test 5
Edited Test 12.

IbrayevRamil committed:
Merge branch 'IR' of https://github.com/IbrayevRamil/javacourses2016 into IR

IbrayevRamil committed:
fixed

nkarulin commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+/.idea/
+/build/
+/.gradle/
```
.gitignore should not be in source repository

nkarulin commented on src/main/java/com/epam/javacourses2016/task1/SolverTask1.java:
```diff
@@ -16,7 +20,32 @@
+            if (!output.exists()) {
+                output.createNewFile();
+            }
+            FileWriter fw = new FileWriter(output.getAbsoluteFile());
```
Why don't you use try-with-resource on Writer? It should also be closed after use

nkarulin commented on src/main/java/com/epam/javacourses2016/task10/SolverTask10.java:
```diff
@@ -17,6 +21,21 @@
+        for(String line: lines){
+            if (words.containsKey(line)) {
+                words.put(line, words.get(line) + 1);
+            }
```
better keep `else` on the same line
```java
if (...) {
   //...
} else {
   //...
}

nkarulin commented on src/main/java/com/epam/javacourses2016/task11/SolverTask11.java:
```diff
@@ -16,16 +16,26 @@
      * @param peoples Список с именами участников.
      * @return Имя последнего оставшегося.
      */
-    public String emulate(ArrayList<String> peoples) {
```
Don't you feel confused about copy-paste in those methods?)

Realization for ArrayList is not effective

nkarulin commented on src/main/java/com/epam/javacourses2016/task12/SolverTask12.java:
```diff
@@ -16,7 +19,30 @@
+                }
+            }
+        }
+        return integers;
```
Can be simplified by using custom Comparator in sorted collection

nkarulin commented on src/main/java/com/epam/javacourses2016/task17/SolverTask17.java:
```diff
@@ -19,7 +19,51 @@
+                if (crossDot != null) {
+                    crossedSeg.put(new Integer[]{i+1,j+1}, crossDot);
+                }
+            }
```
nice but slow:snail: O(n^2)

nkarulin commented on src/main/java/com/epam/javacourses2016/task3/SolverTask3.java:
```diff
@@ -15,8 +16,23 @@
+                    lines.set(j+1, t);
+                }
+            }
+        }
```
Don't use bubble sort in practice:)
Collections.sort is much better

nkarulin commented on src/main/java/com/epam/javacourses2016/task5/SolverTask5.java:
```diff
@@ -18,7 +21,15 @@
+        for (Measurement measurement : measurements) {
+            Sxx = Sxx.add(new BigDecimal(measurement.getCurrent()*measurement.getCurrent()));
+        }
+        BigDecimal R = Syx.divide(Sxx,6);
```
What the magic number?

nkarulin commented on src/main/java/com/epam/javacourses2016/task5/SolverTask5.java:
```diff
@@ -18,7 +21,15 @@
+            Sxx = Sxx.add(new BigDecimal(measurement.getCurrent()*measurement.getCurrent()));
+        }
+        BigDecimal R = Syx.divide(Sxx,6);
+        return R.doubleValue();
```
Are you sure it is method of least squares?

nkarulin commented on src/main/java/com/epam/javacourses2016/task8/SolverTask8.java:
```diff
@@ -16,6 +18,30 @@
+                    return false;
+                }
+                char last = stack.peek();
+                if (current == '}' && last == '{' || current == ')' && last == '(' || current == ']' && last == '[') {
```
can be refactored with enums
Keep data away from code

nkarulin commented on src/main/java/com/epam/javacourses2016/task9/SolverTask9.java:
```diff
@@ -16,7 +20,29 @@
-        return null;
+        List<String> lines = null;
+        try {
+            lines = Files.readAllLines(input.toPath(), StandardCharsets.UTF_8);
```
:snail: 
We dont need that buffer actually, single Set with custom comparator should be enough

nkarulin commented on src/test/java/com/epam/javacourses2016/task1/Task1Test.java:
```diff
@@ -12,7 +12,7 @@
 public class Task1Test {
 
-    @Test(enabled = false, dataProvider = "lines")
+    @Test(enabled = true, dataProvider = "lines")
```
Please enable other tests too

IbrayevRamil committed:
deleted gitignore

IbrayevRamil committed:
all tests enabled, test3 disabled beceause it has problems with
decoding (i dont know why) and because of it fails...i will try to fix this

IbrayevRamil committed:
task13 done

IbrayevRamil committed:
test 3 enabled, i have decoded task3 test files to windows1251, because it
doesnt pass the tests in UTF-8 (idk why), in windows1251 everything is OK,
 task 3 done

IbrayevRamil committed:
из-за смены кодировки, не проходит проверку, возвращаю назад на UTF-8....)

IbrayevRamil committed:
some fixes in task13

IbrayevRamil committed:
task14 done

IbrayevRamil committed:
task16 done

IbrayevRamil committed:
task 15 done

IbrayevRamil commented:
task 15, 16 need some fix

IbrayevRamil committed:
task 15,16 need some fix in reading ad writing to file

IbrayevRamil committed:
task 15 done

IbrayevRamil committed:
task 16 done

IbrayevRamil committed:
some minor fixes, added class matrix for task18

nkarulin commented on src/main/java/com/epam/javacourses2016/task13/Graph.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public void addEdge(int first, int second) {
+        for (Node node : nodes) {
```
Set or Map give us O(logn) or even O(1) (for HashMap or HashSet with no collisions) on searching for element, but your implementations of Graph and Node don't use that speedup using iteration through collection with just O(n) :snail:

nkarulin commented on src/main/java/com/epam/javacourses2016/task13/Node.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Created by Рамиль on 29.11.2016.
+ */
+public class Node {
```
equals/hashCode convention will help you to speed up your graph :rabbit2:

nkarulin commented on src/main/java/com/epam/javacourses2016/task14/MyNumberCollection.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Created by Рамиль on 30.11.2016.
+ */
+public class MyNumberCollection<T extends Number> implements AbstractCollectionCreator.NumberCollection<T> {
```
extending java `AbstractCollection` will help you to save some lines of code

nkarulin commented on src/main/java/com/epam/javacourses2016/task15/IFile.java:
```diff
@@ -0,0 +1,61 @@
+                    final int A_X_LINE_COORDINATE = 0;
+                    final int A_Y_LINE_COORDINATE = 1;
+                    final int B_X_LINE_COORDINATE = 2;
+                    final int B_Y_LINE_COORDINATE = 3;
```
These guys are constants. They should be static final class fields.

nkarulin commented on src/main/java/com/epam/javacourses2016/task16/Cell.java:
```diff
@@ -0,0 +1,45 @@
+    private final Point2D TOP_LEFT_CORNER;
+    private final Point2D BOT_LEFT_CORNER;
+    private final Point2D TOP_RIGHT_CORNER;
+    private final Point2D BOT_RIGHT_CORNER;
```
But these guys are not constants, they are just final fields and should not be written in CAPITAL.

nkarulin commented on src/main/java/com/epam/javacourses2016/task16/Cell.java:
```diff
@@ -0,0 +1,45 @@
+    private final Point2D BOT_LEFT_CORNER;
+    private final Point2D TOP_RIGHT_CORNER;
+    private final Point2D BOT_RIGHT_CORNER;
+    private final double CELL_SIZE = 1.0;
```
This guy is constant it should be static

nkarulin commented on src/main/java/com/epam/javacourses2016/task16/IFile.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Created by Рамиль on 30.11.2016.
+ */
+public class IFile implements SolverTask16.IFileWithPoints {
```
_I_ prefix means _Interface_, actually it is not common for Java more C++ style. But anyway it should not be in naming of classes.

nkarulin commented on src/main/java/com/epam/javacourses2016/task16/IFile.java:
```diff
@@ -0,0 +1,60 @@
+                if (scanner.hasNextDouble()) {
+                    coords.add(scanner.nextDouble());
+                }
+                final int DOT_COORDINATES_AND_DIST = 3;
```
Constant, put it in class header as static

bmkodoo commented:
Nice! But now we need to wait for tests covered all your tasks.
Better keep PR only with covered tasks to merge it earlier.

Also, resolve conflict please

IbrayevRamil committed:
trying to resolve task17 conflict

IbrayevRamil committed:
trying to resolve task17 conflict #2

IbrayevRamil committed:
trying to resolve task17 conflict #3

IbrayevRamil committed:
Merge remote-tracking branch 'remotes/origin/master' into IR

IbrayevRamil committed:
trying to resolve task17 conflict #4, i will enable tests 5 and 17 after fixing my solvers

IbrayevRamil commented:
Finally did it 😁

IbrayevRamil committed:
task 5 fixed, test5 enabled

IbrayevRamil committed:
task17 fixed, test17 enabled

IbrayevRamil committed:
some fixes

Commit:
my tests

bmkodoo commented:
Just noticed. Please change target branch from master to yours.

bmkodoo commented:
master was updated

Commit:
all tests are enabled

nkarulin committed:
Merge pull request #27 from AntonKiselyov/my_tests

1 2 4 5 6 8

IbrayevRamil committed:
tests 11 and 13 disabled. fix in other PR

IbrayevRamil committed:
tests 11 and 13 disabled. fix in other PR

IbrayevRamil committed:
test 11 and 13 fixed

nkarulin committed:
Merge pull request #30 from IbrayevRamil/master

test 11 and 13 fixed

IbrayevRamil committed:
Merge branch 'master' into IR

IbrayevRamil committed:
tasks 1-17 solved

IbrayevRamil committed:
Test 15

IbrayevRamil committed:
Merge branch 'master' into IR

IbrayevRamil committed:
test 11 fix

IbrayevRamil committed:
Merge branch 'master' into IR

IbrayevRamil committed:
test 11 fix #2

IbrayevRamil committed:
Merge branch 'master' into IR

IbrayevRamil committed:
task 11 fix
